File: Leetcode/0094-binary-tree-inorder-traversal/solution.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val; this.left = left; this.right = right;
    }
}

class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode cur = root;
        while (cur != null || !stack.isEmpty()) {
            while (cur != null) { stack.push(cur); cur = cur.left; }
            cur = stack.pop();
            result.add(cur.val);
            cur = cur.right;
        }
        return result;
    }

    static int passed = 0, failed = 0;
    static TreeNode buildTree(Integer[] arr) {
        if (arr.length == 0 || arr[0] == null) return null;
        TreeNode root = new TreeNode(arr[0]);
        Queue<TreeNode> q = new ArrayDeque<>(); q.offer(root);
        int i = 1;
        while (!q.isEmpty() && i < arr.length) {
            TreeNode n = q.poll();
            if (i < arr.length && arr[i] != null) { n.left = new TreeNode(arr[i]); q.offer(n.left); }
            i++;
            if (i < arr.length && arr[i] != null) { n.right = new TreeNode(arr[i]); q.offer(n.right); }
            i++;
        }
        return root;
    }
    static void test(String name, List<Integer> got, List<Integer> want) {
        if (got.equals(want)) { System.out.println("PASS: " + name); passed++; }
        else { System.out.println("FAIL: " + name + " got=" + got); failed++; }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        Object[][] cases = {
            {"Example 1", new Integer[]{1, null, 2, 3}, Arrays.asList(1, 3, 2)},
            {"Empty", new Integer[]{}, Arrays.asList()},
            {"Single", new Integer[]{1}, Arrays.asList(1)},
            {"Left only", new Integer[]{1, 2, null, 3}, Arrays.asList(3, 2, 1)},
            {"Right only", new Integer[]{1, null, 2, null, 3}, Arrays.asList(1, 2, 3)},
            {"Balanced", new Integer[]{1, 2, 3}, Arrays.asList(2, 1, 3)}
        };
        for (Object[] c : cases) test((String) c[0], sol.inorderTraversal(buildTree((Integer[]) c[1])), (List<Integer>) c[2]);
        System.out.printf("%nResults: %d passed, %d failed, %d total%n", passed, failed, passed + failed);
    }
}

File: Leetcode/0094-binary-tree-inorder-traversal/solution.py
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        """Time O(n), Space O(h)."""
        result, stack, cur = [], [], root
        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.left
            cur = stack.pop()
            result.append(cur.val)
            cur = cur.right
        return result


def build_tree(arr):
    if not arr or arr[0] is None: return None
    root = TreeNode(arr[0])
    q = [root]; i = 1
    while q and i < len(arr):
        node = q.pop(0)
        if i < len(arr) and arr[i] is not None:
            node.left = TreeNode(arr[i]); q.append(node.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode(arr[i]); q.append(node.right)
        i += 1
    return root


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    def test(name, got, exp):
        global passed, failed
        if got == exp: print(f"PASS: {name}"); passed += 1
        else: print(f"FAIL: {name} got={got}"); failed += 1
    cases = [
        ("Example 1", [1, None, 2, 3], [1, 3, 2]),
        ("Empty", [], []),
        ("Single", [1], [1]),
        ("Left only", [1, 2, None, 3], [3, 2, 1]),
        ("Right only", [1, None, 2, None, 3], [1, 2, 3]),
        ("Balanced", [1, 2, 3], [2, 1, 3]),
    ]
    for n, t, exp in cases:
        test(n, sol.inorderTraversal(build_tree(t)), exp)
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")
